### ISSUE TITLE
Python-odeint-v2

### ISSUE THREAD

**Issue by 11718pauld** [headmyshoulder/odeint-v2#155]
Is there a wrapper for Python-Odeint-v2 ?

**Comment by mariomulansky** [comment]
Hi,
I'm not aware of any Python bindings for odeint. However, note that there are quiet sophisticated ODE solvers available in the scipy package. 
The main advantage of Boost.odeint is the generality, but this is very difficult to transfer in a python binding so I'm not very convinced Python bindings for odeint make much sense.

**Comment by 11718pauld** [comment]
Well, I was looking for an easy to use GPU based ODE solver.
My system includes 250 sets of Independent systems, where each system posses 350 ODEs.
Hence, the need for GPU based ODE solver !

**Comment by ddemidov** [comment]
It should be easy to wrap the exact solver you need with the exact GPU backend you want into a python module with help of boost.python. Then you could transparently call it from python with numpy arrays as input/output.

**Comment by 11718pauld** [comment]
Why it has to an exact GPU backend? boost is a generalized one !

Well, I am not so expert with c++ and python either !

**Comment by mariomulansky** [comment]
Last time I checked Boost.Python did not support numpy particularly well, but that was quite some ago so maybe it improved.
But I agree, the best would be to write a module in C that does the ODE solving and import in python. I've had good experience with Cython, where I managed to include Boost.odeint. But to get that running with GPU compilers is a different story I suppose.

**Comment by 11718pauld** [comment]
Can you suggest some libraries  (python based) which can solve ODEs on GPU?

**Comment by ddemidov** [comment]
> Why it has to an exact GPU backend? 

Because the solver and backend are selected at compile time, and once you create a python module, you can not change those. You could also compile several options in and let the user choose among them of course. 

> Last time I checked Boost.Python did not support numpy particularly well.

I used it successfully with help of this package: https://github.com/mdboom/numpy-boost. See example here: https://github.com/ddemidov/amgcl/tree/master/pyamgcl.

**Comment by bjodah** [comment]
Sorry if this isn't the appropriate place to post, but in case someone might be interested I wrote this Python wrapper for odeint:

https://github.com/bjodah/pyodeint

It will obviously be quite slow, but I find it useful for trying things out in Python (which is easier than in C++ in my humble opinion). Note that I just got this working so there might be some bugs not caught by the test suite. Any criticism is very welcome.

**Comment by mariomulansky** [comment]
Hi bjodah,
I like your pyodint bindings, it's a nice example how access odeint from Python. Now I'm wondering how much more difficult the interface would become if you somehow allow the rhs function to be defined in Cython as well, because this would then give you a huge performance boost and thus a clear advantage over the exisiting scipy.integrate.odeint imlementation. On the other hand, that would mean the user would have to interact with cython all the time, which decreases usability I suspect...

**Comment by bjodah** [comment]
I actually use this with Cython myself, i.e. I have a rhs function in C++ which is wrapped using Cython. I can then use scipy or pyodeint to call the python callbacks in the Cython generated wrapper. The overhead is actually surprisingly small (given the rhs, and jac calculations involve a large number of operations). For some problems where a very large number of steps are taken (maybe on the order of 10^6) I might suspect a Cython wrapper passing a `cdef` function pointer to be significantly faster, but I guess some benchmarks would be needed to be sure..

**Comment by jackhall** [comment]
Maybe some of you would be interested in a [python library](
https://github.com/jackhall/Lyapunov) I wrote a while ago and maintain occasionally. It's not a straightforward wrapper of odeint, but it does use the odeint solvers. I used boost.python for the language interface.

Pros:
- Rootfinding! (only bisection, but can handle many event functions simultaneously)
- A new and (I think) better abstraction for ODE solving which is geared for ease of use without sacrificing speed, flexibility or numerical guarantees. Pythonic. 
- Good coverage of the explicit solvers. Only missing one or two. It's really easy to add new ones. 
- Documentation. 

Cons:
- No implicit solvers yet. 
- No numpy arrays yet, but adding them shouldn't be too hard. 
- No GPU support or similar fanciness. Not sure how hard this would be. 
- The only tests are a set of demos. 

Let me know if you like it! Nothing like having users to spur development...
